tests/deployments/salt/salt_test.py

```diff
@@ -0,0 +1,119 @@
+import os
+import tempfile
+from functools import partial as p
+from pathlib import Path
+
+import pytest
+import yaml
+
+from tests.helpers.assertions import has_datapoint_with_dim, has_datapoint_with_metric_name
+from tests.helpers.util import print_lines, wait_for, copy_file_into_container
+from tests.packaging.common import (
+    INIT_SYSTEMD,
+    get_agent_logs,
+    get_agent_version,
+    is_agent_running_as_non_root,
+    run_init_system_image,
+)
+from tests.paths import REPO_ROOT_DIR
+
+pytestmark = [pytest.mark.salt, pytest.mark.deployment]
+
+DOCKERFILES_DIR = Path(__file__).parent.joinpath("images").resolve()
+
+DEB_DISTROS = [
+    ("debian-8-jessie", INIT_SYSTEMD),
+    ("debian-9-stretch", INIT_SYSTEMD),
+    ("ubuntu1604", INIT_SYSTEMD),
+    ("ubuntu1804", INIT_SYSTEMD),
+]
+
+RPM_DISTROS = [("amazonlinux2", INIT_SYSTEMD), ("centos7", INIT_SYSTEMD), ("centos8", INIT_SYSTEMD)]
+
+CONFIG = """
+signalfx-agent:
+  package_stage: null
+  version: null
+  conf:
+    signalFxAccessToken: 'testing123'
+    ingestUrl: null
+    apiUrl: null
+    intervalSeconds: 1
+    observers:
+      - type: host
+    monitors: null
+"""
+
+PILLAR_PATH = "/srv/pillar/signalfx-agent.sls"
+SALT_CMD = "salt-call --local state.apply"
+STAGE = os.environ.get("STAGE", "final")
+INITIAL_VERSION = os.environ.get("INITIAL_VERSION", "4.7.5-1")
+UPGRADE_VERSION = os.environ.get("UPGRADE_VERSION", "4.7.6-1")
+
+
+def get_config(backend, agent_version, monitors, stage):
+    config_yaml = yaml.safe_load(CONFIG)
+    config_yaml["signalfx-agent"]["package_stage"] = stage
+    config_yaml["signalfx-agent"]["version"] = agent_version
+    config_yaml["signalfx-agent"]["conf"]["ingestUrl"] = backend.ingest_url
+    config_yaml["signalfx-agent"]["conf"]["apiUrl"] = backend.api_url
+    config_yaml["signalfx-agent"]["conf"]["monitors"] = monitors
+    return yaml.dump(config_yaml)
+
+
+def run_salt(cont, backend, agent_version, monitors, stage):
+    with tempfile.NamedTemporaryFile(mode="w+") as fd:
```

Comment:
Add some blank lines between "paragraphs" of code -- it's kind of hard to read.  I would start by putting each assertion or set of adjacent assertions surrounded by blank lines to make them stand out.